substrate/frame/mixnet/src/lib.rs

```diff
@@ -0,0 +1,598 @@
+		/// transactions.
+		#[pallet::constant]
+		type NumRegisterStartSlackBlocks: Get<BlockNumberFor<Self>>;
+
+		/// The number of "slack" blocks at the end of each session.
+		/// [`maybe_register`](Pallet::maybe_register) will try to register before this slack
+		/// period, but may post registration transactions during the slack period as a last
+		/// resort.
+		#[pallet::constant]
+		type NumRegisterEndSlackBlocks: Get<BlockNumberFor<Self>>;
+
+		/// Priority of unsigned transactions used to register mixnodes.
+		#[pallet::constant]
+		type RegistrationPriority: Get<TransactionPriority>;
+
+		/// Minimum number of mixnodes. If there are fewer than this many mixnodes registered for a
+		/// session, the mixnet will not be active during the session.
+		#[pallet::constant]
+		type MinMixnodes: Get<u32>;
+	}
+
+	/// Index of the current session. This may be offset relative to the session index tracked by
+	/// eg `pallet_session`; mixnet session indices are independent.
+	#[pallet::storage]
+	pub(crate) type CurrentSessionIndex<T> = StorageValue<_, SessionIndex, ValueQuery>;
+
+	/// Block in which the current session started.
+	#[pallet::storage]
+	pub(crate) type CurrentSessionStartBlock<T> = StorageValue<_, BlockNumberFor<T>, ValueQuery>;
+
+	/// Authority list for the next session.
+	#[pallet::storage]
+	pub(crate) type NextAuthorityIds<T> = StorageMap<_, Identity, AuthorityIndex, AuthorityId>;
+
+	/// Mixnode sets by session index. Only the mixnode sets for the previous, current, and next
+	/// sessions are kept; older sets are discarded.
+	///
+	/// The mixnodes in each set are keyed by authority index so we can easily check if an
+	/// authority has registered a mixnode. The authority indices should only be used during
+	/// registration; the authority indices for the very first session are made up.
+	#[pallet::storage]
+	pub(crate) type Mixnodes<T> =
+		StorageDoubleMap<_, Identity, SessionIndex, Identity, AuthorityIndex, BoundedMixnodeFor<T>>;
+
+	#[pallet::genesis_config]
+	#[derive(frame_support::DefaultNoBound)]
+	pub struct GenesisConfig<T: Config> {
+		/// The mixnode set for the very first session.
+		pub mixnodes: BoundedVec<BoundedMixnodeFor<T>, T::MaxAuthorities>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			assert!(
+				Mixnodes::<T>::iter_prefix_values(0).next().is_none(),
+				"Initial mixnodes already set"
+			);
+			for (i, mixnode) in self.mixnodes.iter().enumerate() {
+				// We just make up authority indices here. This doesn't matter as authority indices
+				// are only used during registration to check an authority doesn't register twice.
+				Mixnodes::<T>::insert(0, i as AuthorityIndex, mixnode);
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register a mixnode for the following session.
+		#[pallet::call_index(0)]
+		#[pallet::weight(1)] // TODO
```

Comment:
Does this need a benchmark?

Reply:
Yes, but this pallet doesn't go to production directly?

Reply:
Sure, can be done as a follow-up, but our policy is to use tracking issues instead of TODOs :P

Reply:
Yeah, fair point.